File: leetcode/19.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode* next;
 * };
 */
struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    int len = 0;
    struct ListNode* curr = head;
    while (curr != NULL) {
        len++;
        curr=curr->next;
    }
    if (len == n) return head->next;
    curr = head;
    while (len > n + 1) {
        len--;
        curr=curr->next;
    }
    curr->next = curr->next->next;
    return head;
}

File: leetcode/19.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # if head.next is None: return None
        # if head.next.next is None:
        #     if n == 2: return head.next
        #     head.next = None
        #     return head

        frnt = head
        back = head
        diff = 0

        while frnt.next is not None:
            frnt = frnt.next
            diff = diff + 1
            if diff > n:
                back = back.next
                diff = diff - 1

        if diff < n: return head.next

        back.next = back.next.next
        return head
